canonical-web-and-design/maas-ui
# fix(ui): fix bug where boot resources that share os/release/arch were handled incorrectly

## Done

- Include the boot resource id in the form image values, instead of matching based on identical os/release/arch.

## QA

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Instead of downloading images to setup the MAAS properly, we will just fake the data in the store. Open `ui/src/app/store/bootresource/selectors.ts` and add:
 ``` ts
  import type { BootResource } from "./types";
  import { bootResource as bootResourceFactory } from "testing/factories";
  ```
- Replace the `uploadedResources` selector on line 73 with some boot resources that have identical os + release + arch:
  ``` ts
  const uploadedResources = (_: RootState): BootResource[] => [
    bootResourceFactory({ id: 123, name: "os/release", arch: "amd64", title: "Title 1" }),
    bootResourceFactory({ id: 456, name: "os/release", arch: "amd64", title: "Title 2" }),
  ];
  ```
- Running the `master` branch, navigate to the Image list and check that there is a Custom images section, but both of them have the same title
- Click the bin icon for one of them and check that both rows expand with the delete confirmation
- Switch to this branch and check that both titles are represented and clicking the bin icon only expands the row that you clicked on
- Check that you can still add, clear, delete and save Ubuntu/other images as before.

## Fixes

Fixes #3266 

## Launchpad issue

[lp#1950694](https://bugs.launchpad.net/maas/+bug/1950694)
